### PR TITLE
Implement `autobib update`, `autobib info`, and `autobib delete`

### PR DESCRIPTION
**Pull request by alexrutar**
Stuff that has to be done:
- [x] Implement database functions
  - [x] for `delete`
  - [x] for `update`
- [x] Implement "associated keys" to get everything in citation keys points to a given record
  - [x] Also use this to implement `autobib info`
- [x] Prompt user for confirmation before deleting
  - [x] Create new 'confirm' module in `term.rs`, similar to edit, and also use this in the `edit` module
  - [x] Add option to skip confirmation prompt
- [x] Implement `get_remote_data` function, to convert a `RemoteId` into a `RawRecordData`.
- [x] Prompt user for confirmation before deleting if there are multiple keys.

### CODE REVIEW

**Comment by alexrutar** [comment]
Well, everything is implemented; unfortunately, the code quality is a bit of a disaster.

File: src/db/data.rs [hunk 1]
```diff
@@ -216,20 +216,27 @@ impl RawRecordData {
...
-        let (type_block, _) = self.split_blocks();
-        from_utf8(&type_block[1..]).unwrap()
-    }
+            fn entry_type(&self) -> &str {
+                let (type_block, _) = self.split_blocks();
+                from_utf8(&type_block[1..]).unwrap()
+            }
+        }
+    };
 }
```
Comment (wupr) [inline src/db/data.rs:235]: We can explicitly disallow an empty series of tokens.
```suggestion
macro_rules! entry_data_impl {
    ($($raw:tt)+) => {
        impl EntryData for $($raw)+ {
            fn fields(&self) -> impl Iterator<Item = (&str, &str)> {
                let (_, data_blocks) = self.split_blocks();
                RawRecordFieldsIter {
                    remaining: data_blocks,
                }
            }

            fn entry_type(&self) -> &str {
                let (type_block, _) = self.split_blocks();
                from_utf8(&type_block[1..]).unwrap()
            }
        }
    };
}
```

**Comment by wupr** [comment]
> Well, everything is implemented; unfortunately, the code quality is a bit of a disaster.

Nice. Should I review now or do you still want to make changes?

**Comment by alexrutar** [comment]
I hope to clean up the implementation a bit, but I currently am very undecided on what type of `Transaction`-type abstraction the database should offer! I will let you know / request a review when the PR is in a good state.

**Comment by wupr** [comment]
Cool! I guess I'll just try running a few commands for now.

**Comment by alexrutar** [comment]
There are still many things not implemented; in particular the changes discussed in #91 #90 #89 #72 

However, this PR is already so large that I hesitate to make more changes here, until it is merged.

File: src/db/row.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+/// A Wrapper type to represent a row in the `Records` table which either exists
+/// or is missing.
+#[derive(Debug)]
+pub enum DatabaseEntry<'conn> {
+    /// The row exists.
+    Exists(Row<'conn>),
+    /// The row is missing.
+    Missing(Missing<'conn>),
+}
```
Comment (wupr) [inline src/db/row.rs:18]: How about renaming `Row` to `RecordRow` and `Missing` to `MissingRecordRow`?

**Comment by wupr** [comment]
I haven't completely read through `row.rs`, but why don't the operations like `delete_row_data` and `get_referencing_keys` just take `RecordRow` as an argument? That would ensure consistency between the transaction and the row id. Similarly, `check_null` can take `MissingRecordRow`.

**Comment by wupr** [comment]
Actually, would it make sense to combine the two types into a single type that contains a reference to transaction and an optional `row_id`?

**Comment by alexrutar** [comment]
Hmm actually we could even just implement these as methods on the `(Missing)RecordRow`s directly?

**Comment by alexrutar** [comment]
My initial motivation was to allow the caller to specify the function, without needing to access internal state of the `RecordRow`. But I guess in practice, we can just implement whatever functions we want.

**Comment by alexrutar** [comment]
Either way, I would agree though that just passing the `RecordRow` directly, and providing methods to access the internal state, would be ideal.

**Comment by wupr** [comment]
Actually, yeah, I would support implementing the methods on the instances directly.

I think I have understood the paradigm you were proposing and it looks good to me.

**Comment by wupr** [comment]
Would it be useful to keep track of whether the transaction needs committing (i.e. whether the database is being modified)?

**Comment by alexrutar** [comment]
Wait one problem with requiring the methods to be implemented directly on the database is that it interferes with composition. Let's say I am a caller, and I want to define a custom operation which say does `A -> B -> C -> D`. If we don't keep the current `apply`-style implementation, then this custom operation needs to either be (1) applied inline whenever it is needed, or (2) implemented as a method on the `(Missing)RecordRow` directly.

On the other hand, if I want to define `my_custom_method`, which just executes `A -> B -> C -> D`, I can easily do this myself, which helps to separate "application logic" and "database logic".

One the issue with the previous implementation was that all of the logic required by every caller needed to be implemented directly in the `db.rs`, which make the `RecordDatabase` implementation extremely unreadable.

**Comment by alexrutar** [comment]
Anyway, the point is that in the `db` module, implement all of the "core" operations, and then whenever some sort of compound operation is required by the caller, they can wrap together core operations themselves.

**Comment by alexrutar** [comment]
> Would it be useful to keep track of whether the transaction needs committing (i.e. whether the database is being modified)?

Good question, it's possible. Ideally this would be a compile-time check I guess.

**Comment by alexrutar** [comment]
I pushed some changes which converts the naming to `(Missing)RecordRow`. Also, now the signatures are changed to require a `(Missing)RecordRow`.

**Comment by alexrutar** [comment]
I also moved the `operations` into a different module.

**Comment by wupr** [comment]
> Wait one problem with requiring the methods to be implemented directly on the database is that it interferes with composition. Let's say I am a caller, and I want to define a custom operation which say does `A -> B -> C -> D`. If we don't keep the current `apply`-style implementation, then this custom operation needs to either be (1) applied inline whenever it is needed, or (2) implemented as a method on the `(Missing)RecordRow` directly.

Without `apply`, one could still write a function `do_abc` that takes `row` and return `row.do_a().do_b().do_c()`, and whenever needed call `row = do_abc(row)`. But perhaps this is not as nice as `row.do_abc()`.

I think it's fine to keep `apply`, but it feels arbitrary to me that `delete` is in `impl<'conn> RecordRow<'conn>` but `update_row_data` isn't.

**Pull request merged**